chore: Document memory management in Grain (#768)
grain-lang/grain: docs/contributor/memory_management.md

# Memory Management in Grain

This guide documents Grain's memory management system. This document is aimed at those who are developing the code-generation pass of the compiler or writing low-level code inside of the Grain runtime; most users should not need to interact with the memory manager, and doing so risks corrupting the memory of your program (if the guidelines laid out here are not adhered to).

We ultimately aim to replace Grain's in-house memory management with the WebAssembly GC instructions once [the proposal](https://github.com/WebAssembly/gc/blob/master/proposals/gc/Overview.md) is implemented, but, in the meantime, Grain implements its own memory allocator and garbage collector.

## Memory Allocator

> Grain uses a [memory allocator](https://github.com/grain-lang/grain/blob/main/stdlib/runtime/malloc.gr) derived from the `malloc`/`free` example given in Kernighan and Ritchie's "The C Programming Language" (K&R C). This module exports the following values:

Reviewer: Might as well link the book 😂 
```suggestion
Grain uses a [memory allocator](https://github.com/grain-lang/grain/blob/main/stdlib/runtime/malloc.gr) derived from the `malloc`/`free` example given in Kernighan and Ritchie's ["The C Programming Language"](https://kremlin.cc/k&r.pdf) (K&R C), pages 185-188 (PDF page 199). This module exports the following values:
```

 * @return WasmI32 - The given pointer
 */
export let incRef = (userPtr: WasmI32) => WasmI32

/**
 * Decrements the reference count of the given pointer if it is greater than zero
 *
 * @param userPtr: WasmI32 - The pointer whose reference count should be decremented
 * @return WasmI32 - The given pointer
 */
export let decRefIgnoreZeros = (userPtr: WasmI32) => WasmI32

/**
 * Decrements the reference count of the given pointer. An error is thrown if the
 * reference count is not greater than zero.
 *
 * @param userPtr: WasmI32 - The pointer whose reference count should be decremented
 * @return WasmI32 - The given pointer
 */
export let decRef = (userPtr: WasmI32) => WasmI32
```

When writing code in the runtime, `decRefIgnoreZeros` should never be used. It exists solely to support the cleanup step of compiled functions.

### Garbage Collection in Compiled Programs

In most Grain programs, the above functions do not need to be invoked by users. Instead, the compiler automatically inserts calls to them where required. When developing in `compcore.re`, it is important to understand the conventions used by the garbage collector.

> **References from the local frame still count as references.** In other words, when a variable is stored inside of a WASM local, its reference count should be incremented. There is one exception to this rule: so-called "swap slots". These are special locals that Grain uses to store temporary results within a single compiled `Mashtree` instruction (similar to how CPU registers are used in native compilers). For performance reasons, we do not increment the reference counter when storing values in these slots. This is safe due to the fact that these values do not escape beyond a single `Mashtree` instruction; while they may still be physically present in the local, they are not read by later pieces of the program.

Reviewer: ```suggestion
**References from the local frame still count as references.** In other words, when a variable is stored inside of a Wasm local, its reference count should be incremented. There is one exception to this rule: so-called "swap slots". These are special locals that Grain uses to store temporary results within a single compiled `Mashtree` instruction (similar to how CPU registers are used in native compilers). For performance reasons, we do not increment the reference counter when storing values in these slots. This is safe due to the fact that these values do not escape beyond a single `Mashtree` instruction; while they may still be physically present in the local, they are not read by later pieces of the program.
```

We ultimately aim to replace Grain's in-house memory management with the WebAssembly GC instructions once [the proposal](https://github.com/WebAssembly/gc/blob/master/proposals/gc/Overview.md) is implemented, but, in the meantime, Grain implements its own memory allocator and garbage collector.

## Memory Allocator
Grain uses a [memory allocator](https://github.com/grain-lang/grain/blob/main/stdlib/runtime/malloc.gr) derived from the `malloc`/`free` example given in Kernighan and Ritchie's ["The C Programming Language"](https://kremlin.cc/k&r.pdf) (K&R C), pages 185-188 (PDF page 199). This module exports the following values:

```grain
/**
 * The amount of space reserved for runtime code which needs to be executed
 * at boot before the malloc module is loaded
 */
export let _RESERVED_RUNTIME_SPACE: WasmI32

/**
 * Allocates the requested number of bytes, returning a pointer.
 *
 * @param nbytes: WasmI32 - The number of bytes to allocate
 * @return WasmI32 - The pointer to the allocated region (8-byte aligned), or -1 if the allocation failed.
 */

> export let malloc: (nbytes: WasmI32) => WasmI32

Reviewer 2: Do we use `->` for signatures?
```suggestion
export let malloc: (nbytes: WasmI32) -> WasmI32
```

```grain
let f = x => {
  incRef(x)
  foo(x)
  decRef(x)
  incRef(x)
  bar(x)
  decRef(x)
}
```
Grain's compiler will emit something more like this (again, note that `x` is an argument, so it is incumbent upon `f`'s caller to `decRef` it):
```grain
let f = x => {
  foo(x)
  bar(x)
}
```

## Disabling the Garbage Collector

There are times in which the garbage collector is not desired in Grain code. The typical example of this is the implementation of the core runtime (for example the garbage collector itself), which involves working with native WebAssembly values. Consequently, Grain provides two mechanisms for disabling the garbage collector, either on a per-function or a per-file basis.

First, it is possible to annotate functions with `@disableGC`. This will compile the annotated function with no garbage collection instructions. Second, it is possible to pass the `--compilation-mode=runtime` flag to the compiler (**NOTE** that this is not the only effect of this flag; see `runtime.md` for more details), which is typically done by placing this at the beginning of the file:
```grain

> /* grainc-flags --compilation-mode=runtime */

Reviewer 2: I thought runtime mode has GC but no pervasives? Though there is a `--disable-gc` flag 🤔

Reviewer: Runtime mode doesn't have GC (since it implements GC). We managed to get rid of `--compilation-mode=managed-runtime` which had that behavior, since you can just use `--no-pervasives` now.

Reviewer 2: My mistake 🤦

```
swap_slot <- return_value
for argument in arguments:
    decref(argument)
for local_variable in local_variables:
    decref(local_variable)
return load_swap(return_value) # see note below
```

Note that `load_swap` is one of the aforementioned exceptions which does not `incRef` the value upon access. Swap slots are used by the compiler to temporarily store intermediate values during computation, and they do not need to have their contents `incRef`ed; i.e. since they are only used for intermediate values, any references they hold are treated as weak and do not impact the reference count.

**References are callee-owned.** This has a couple of implications:

- If a function is called with an argument, it is the responsibility of that function to `decRef` that argument
- If a value is passed to a function with a reference count of `n`, when it returns, it will have the a reference count of `n-1` (unless it was stored in an external location, in which case the reference count may be higher).

> **Functions contain references to themselves.** In order to support closures, Grain's compiler emits code which passes the current closure to each function as an extra argument. With respect to the calling convention, this extra argument is the same as normal arguments, so it follows the same `incRef`-before-function-call pattern as standard arguments (and must be `decRef`ed before functions return). This is particularly important to keep in mind when writing `@disableGC` code, as discussed below.

Reviewer: I'd love to see this info before the `Reference counts are incremented upon access` section! I like having the concepts and then getting into our implementation of it. It might even be better to have two subsections of implementations—`Reference counts are incremented upon access` for how it's implemented in the compiler, and then this other bit below for how it should be implemented in Grain code 😎

Author: @ospencer I did the rearrangement, but I am not sure what would be the best way to break this apart into subsections as you describe. Do you want to take another look and see if it now looks like what you'd expect?

Reviewer: Looks great!

## Disabling the Garbage Collector

There are times in which the garbage collector is not desired in Grain code. The typical example of this is the implementation of the core runtime (for example the garbage collector itself), which involves working with native WebAssembly values. Consequently, Grain provides two mechanisms for disabling the garbage collector, either on a per-function or a per-file basis.

> First, it is possible to annotate functions with `@disableGC`. This will compile the annotated function with no garbage collection instructions. Second, it is possible to pass the `--compilation-mode=runtime` flag to the compiler (**NOTE** that this is not the only effect of this flag; see `runtime.md` for more details), which is typically done by placing this at the beginning of the file:

Reviewer: This isn't quite right—the flag to turn off GC at the file level is `--no-gc`. `--compilation-mode=runtime` doesn't emit GC instructions more because the GC doesn't exist rather than it being disabled. We can probably just mention that code written in runtime mode that interacts with user code needs manual management.

If we wanted we could also include some information like this:
`--no-gc`: disables the garbage collector, allocations are still done via `malloc`
`--compilation-mode=runtime`: GC doesn't exist, implicit allocations can't be reclaimed

# Memory Management in Grain

> This guide documents Grain's memory management system. This document is aimed at those who are developing the code-generation pass of the compiler or writing low-level code inside of the Grain runtime; most users should not need to interact with the memory manager, and doing so risks corrupting the memory of your program (if the guidelines laid out here are not adhered to).

Reviewer 2: ```suggestion
This guide documents Grain's memory management system. This document is aimed at those who are developing the code-generation pass of the compiler or writing low-level code inside of the Grain runtime; most users should not need to interact with the memory manager, and doing so risks corrupting the memory of your program if the guidelines laid out here are not adhered to.
```

> We ultimately aim to replace Grain's in-house memory management with the WebAssembly GC instructions once [the proposal](https://github.com/WebAssembly/gc/blob/master/proposals/gc/Overview.md) is implemented, but, in the meantime, Grain implements its own memory allocator and garbage collector.

Reviewer 2: ```suggestion
We ultimately aim to replace Grain's bespoke memory management with the WebAssembly GC instructions once [the proposal](https://github.com/WebAssembly/gc/blob/master/proposals/gc/Overview.md) is implemented—in the meantime, Grain implements its own memory allocator and garbage collector.
```
😛

## Memory Allocator

Grain uses a [memory allocator](https://github.com/grain-lang/grain/blob/main/stdlib/runtime/malloc.gr) derived from the `malloc`/`free` example given in Kernighan and Ritchie's ["The C Programming Language"](https://kremlin.cc/k&r.pdf) (K&R C), pages 185-188 (PDF page 199). This module exports the following values:

```grain
/**
 * The amount of space reserved for runtime code which needs to be executed
 * at boot before the malloc module is loaded
 */
export let _RESERVED_RUNTIME_SPACE: WasmI32

/**
 * Allocates the requested number of bytes, returning a pointer.
 *

>  * @param nbytes: WasmI32 - The number of bytes to allocate
>  * @return WasmI32 - The pointer to the allocated region (8-byte aligned), or -1 if the allocation failed.

Reviewer 2: ```suggestion
 * @param nbytes: The number of bytes to allocate
 * @returns The pointer to the allocated region (8-byte aligned), or -1 if the allocation failed.
```
Might as well follow graindoc?

 */
export let malloc: (nbytes: WasmI32) -> WasmI32

/**
 * Frees the given allocated pointer.
 *

>  * @param ap: WasmI32 - The pointer to free

Reviewer 2: ```suggestion
 * @param ap: The pointer to free
```
Might as well follow graindoc?

 */
export let free = (ap: WasmI32) => Void

/**

>  * Returns the current free list pointer
>  * (used for debugging)

Reviewer 2: ```suggestion
 * Returns the current free list pointer (used for debugging)
```

 *

>  * @return WasmI32 - The free list pointer

Reviewer 2: ```suggestion
 * @returns The free list pointer
```
Might as well follow graindoc?

 */
export let getFreePtr = () => WasmI32
```

These functions should be familiar to programmers who have used `malloc` and `free` in C (and C-like languages). For further reading, refer to this Wikipedia page: [C dynamic memory allocation](https://en.wikipedia.org/wiki/C_dynamic_memory_allocation). The semantics of these functions align near-identically with those of C's corresponding functions.

> These functions generally should not be called directly in 99% of circumstances, and the `GC` module's wrappers should be used instead (described below).

Reviewer 2: ```suggestion
These functions generally should not be called directly in 99% of circumstances, and the Garbage Collector wrappers should be used instead (described below).
```

Author: You sure about this one? I thought "`GC` module" would be better at making it explicit _where_ these wrappers are.

Reviewer 2: I'm fine if we keep the current language.

## Garbage Collector

> Functional programming patterns often result in a large number of allocations, so it is imperative for developer experience that the language manages the creation and deletion of objects for users. Grain uses a reference-counting [garbage collector](https://github.com/grain-lang/grain/blob/main/stdlib/runtime/gc.gr) in order to determine when objects are no longer in active use by the program (and therefore able to have their memory freed).

Reviewer 2: ```suggestion
Functional programming patterns often result in a large number of allocations, so it is imperative for developer experience that the language manages the creation and deletion of objects for users. Grain uses a reference-counting [garbage collector](https://github.com/grain-lang/grain/blob/main/stdlib/runtime/gc.gr) in order to determine when objects are no longer in active use by the program and therefore able to have their memory freed.
```
Just trying to remove parens

At a technical level, this is represented in-memory in the following way, for an n-byte heap object:

```
 [ 32-bit counter ][ 32-bit padding ][ n-byte payload ]
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
 {start address}                    {pointer used by Grain runtime}
```

Here, `{start address}` is the value returned by `Malloc.malloc` (and passed to `Malloc.free` when the object is cleaned up), and the `{pointer used by Grain runtime}` is what everything outside of the `GC` wrapper uses as the pointer to the object. In other words, we "hide" the reference count in the 8 bytes prior to the pointer.

The interface provided by the `GC` module is similar (but not identical) to that which is provided by the `Malloc` module:

```grain
/**
 * Allocates the requested number of bytes, returning a pointer with a reference count of 1.
 *

>  * @param size: WasmI32 - The number of bytes to allocate
>  * @return WasmI32 - The pointer to the allocated region

Reviewer 2: ```suggestion
 * @param size: The number of bytes to allocate
 * @returns The pointer to the allocated region
```
Might as well follow graindoc?

 */
export let malloc = (size: WasmI32) -> WasmI32

/**
 * Frees the given pointer. Using this pointer after it has been freed will result in undefined behavior.
 *

>  * @param userPtr: WasmI32 - The pointer to free
>  * @return Void

Reviewer 2: ```suggestion
 * @param userPtr: The pointer to free
```
Might as well follow graindoc?

 */
export let free = (userPtr: WasmI32) -> Void

/**
 * Increments the reference count of the given pointer.
 *

>  * @param userPtr: WasmI32 - The pointer whose reference count should be incremented
>  * @return WasmI32 - The given pointer

Reviewer 2: ```suggestion
 * @param userPtr: The pointer whose reference count should be incremented
 * @returns The given pointer
```
Might as well follow graindoc?

 */
export let incRef = (userPtr: WasmI32) -> WasmI32

/**
 * Decrements the reference count of the given pointer if it is greater than zero
 *

>  * @param userPtr: WasmI32 - The pointer whose reference count should be decremented
>  * @return WasmI32 - The given pointer

Reviewer 2: ```suggestion
 * @param userPtr: The pointer whose reference count should be decremented
 * @returns The given pointer
```
Might as well follow graindoc?

 */
export let decRefIgnoreZeros = (userPtr: WasmI32) -> WasmI32

/**
 * Decrements the reference count of the given pointer. An error is thrown if the
 * reference count is not greater than zero.
 *

>  * @param userPtr: WasmI32 - The pointer whose reference count should be decremented
>  * @return WasmI32 - The given pointer

Reviewer 2: ```suggestion
 * @param userPtr: The pointer whose reference count should be decremented
 * @returns The given pointer
```
Might as well follow graindoc?

 */
export let decRef = (userPtr: WasmI32) -> WasmI32
```

When writing code in the runtime, `decRefIgnoreZeros` should never be used. It exists solely to support the cleanup step of compiled functions. The reference count-managing functions are safe to use with non-pointers; if a non-pointer is passed

> to them, it will be returned without any side effects. **However**, it should be noted that these functions are not safe to use with arbitrary (untagged) `WasmXX` (`WasmI32`/`WasmI64`/etc) values, which is why the use of such types is only permitted in `@disableGC` blocks (so that the compiler does not insert `incRef`/`decRef` calls).

Reviewer 2: ```suggestion
to them, it will be returned without any side effects. **However**, it should be noted that these functions are not safe to use with arbitrary (untagged) `WasmXX` (`WasmI32`/`WasmI64`/etc) values, which is why the use of such types is only permitted in `@disableGC` blocks—so that the compiler does not insert `incRef`/`decRef` calls.
```

### Garbage Collection in Compiled Programs

> In most Grain programs, the above functions do not need to be invoked by users. Instead, the compiler automatically inserts calls to them where required. When developing in `compcore.re`, it is important to understand the conventions used by the garbage collector.

Reviewer: Want to remove any thought that users sometimes need to do it:
```suggestion
In normal Grain programs, the above functions do not need to be invoked by users. Instead, the compiler automatically inserts calls to them where required. When developing in `compcore.re`, it is important to understand the conventions used by the garbage collector.
```

> **Functions contain references to themselves.** In order to support closures, Grain's compiler emits code which passes the current closure to each function as an extra argument. With respect to the calling convention, this extra argument is the same as normal arguments, so it follows the same `incRef`-before-function-call pattern as standard arguments (and must be `decRef`ed before functions return), as described in the following section. This is particularly important to keep in mind when writing `@disableGC` code, as discussed in "Disabling the Garbage Collector" below.

Reviewer 2: ```suggestion
**Functions contain references to themselves.** In order to support closures, Grain's compiler emits code which passes the current closure to each function as an extra argument. With respect to the calling convention, this extra argument is the same as normal arguments, so it follows the same `incRef`-before-function-call pattern as standard arguments and must be `decRef`ed before functions return, as described in the following section. This is particularly important to keep in mind when writing `@disableGC` code, as discussed in "Disabling the Garbage Collector" below.
```

**Reference counts are incremented upon access.** With some exceptions in generated code, reference counts are incremented whenever a value is accessed (loaded from storage slots such as arguments, locals, and globals). This streamlines the conventions in generated code, as function calls require no specific `incRef` invocation for arguments (see "References are callee-owned" subsection below). For clarity, a couple of notes:

> - If a value is stored somewhere, no extra `incRef`/`decRef` calls are needed (again, there are rare exceptions to this in the compiler, specifically regarding "swap slots", which are local variables used as registers).

Reviewer 2: ```suggestion
- If a value is stored somewhere, no extra `incRef`/`decRef` calls are needed. There are rare exceptions to this in the compiler, specifically regarding "swap slots", which are local variables used as registers.
```

- If the computation no longer needs a live value (e.g. after a statement in the middle of a block), it should be `decRef`ed.
- No special `incRef` is needed for return values, as the access of those values performs the requisite `incRef`. It should be noted that in generated WebAssembly code, the function postamble looks as follows:

```
swap_slot <- return_value
for argument in arguments:
    decref(argument)
for local_variable in local_variables:
    decref(local_variable)
return load_swap(return_value) # see note below
```

Note that `load_swap` is one of the aforementioned exceptions which does not `incRef` the value upon access. Swap slots are used by the compiler to temporarily store intermediate values during computation, and they do not need to have their contents `incRef`ed; i.e. since they are only used for intermediate values, any references they hold are treated as weak and do not impact the reference count.

**References are callee-owned.** This has a couple of implications:

> - If a function is called with an argument, it is the responsibility of that function to `decRef` that argument

Reviewer 2: ```suggestion
- If a function is called with an argument, it is the responsibility of that function to `decRef` that argument.
```
nit
